### ISSUE TITLE
neighborlist in non-orthorhombic systems

### ISSUE THREAD

**Issue by sirmarcel** [google/jax-md#223]
Hello,

This issue is mainly to start a discussion on the interface/behaviour of `neighbor_list` in the case of non-orthorhombic systems, i.e. when the lattice vectors in the `box` are not orthogonal. In that case, the `box_size` needs to be modified (it cannot be the lengths of the lattice vectors), otherwise, wrong results are returned. 

I've written a more detailed explanation [here](https://colab.research.google.com/drive/1BT68skpkahddRODV5kR5un7x25FGnEPI?usp=sharing), but the short version is that in non-orthorhombic systems, the bins in fractional coordinates correspond to parallelpepipeds in real space, so the real space cutoff is given by the distance between opposing faces, rather than the lengths of the edges.

I think this may also cause trouble with updating the neighborlist, but I haven't tested this.

In any case, I would suggest at least clarifying the docs for `box_size`. It may also be useful to introducing some helper/wrapper that deals with the (probably not uncommon) case of running a real space simulation with a fractional neighborlist.

**Comment by sschoenholz** [comment]
Yes! You're totally right that this is (err, was?) a significant problem with neighbor lists in triclinic systems. I hope the situation should be slightly improved following this commit: 55817f84899b0907e285c65afc50dccf41e7bcc4 which uses the distance between opposing faces to produce cell lists with fractional coordinates. Currently, it only works with upper-triangular matrices, but at least it should unblock this case. It also enables slightly better error reporting, which should allow users to identify the need to recompile when the box changes to much with e.g. NPT simulations.

Let me know whether you think this is a step in the right direction or if you have any comments / changes you'd like to see.

**Comment by sirmarcel** [comment]
Hi Sam, thanks! This looks like a solid preliminary fix. 

One thing I haven't tested yet, but am worried about is how changes in the `box` affect the decision on whether to update the neighborlist. Changing the `box` changes the implicit cutoff defined by the cell list bins, and so (I think?) one has to take that change into account in the `jnp.any(d(position, nbrs.reference_position) > threshold_sq)` call, for instance by updating the threshold. If the cell expands, you basically gain additional tolerance, but if it contracts, it shrinks. When it hits the exact threshold, one needs to reallocate, since you start missing neighbours once it goes below the threshold.

I haven't fully worked this out yet -- I don't think `jax-md` necessarily needs to support this right now, once I have some time I will experiment with this some more, but we should probably think about what can go wrong in NPT and make sure warnings are emitted.

**Comment by sirmarcel** [comment]
Btw, on a tangent, it may be a good idea to document the convention of `cell` columns being the lattice vectors, to avoid issues like #220. `ase` uses the opposite convention, for instance.
